File: 0017.letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> result;
    unordered_map<char,vector<char>> um;
    vector<string> letterCombinations(string digits) {
        if(digits.size()==0) return result;
        um['2']=vector<char>({'a','b','c'});
        um['3']=vector<char>({'d','e','f'});
        um['4']=vector<char>({'g','h','i'});
        um['5']=vector<char>({'j','k','l'});
        um['6']=vector<char>({'m','n','o'});
        um['7']=vector<char>({'p','q','r','s'});
        um['8']=vector<char>({'t','u','v'});
        um['9']=vector<char>({'w','x','y','z'});
        string curr="";
        helper(digits,curr,0);
        return result;
    }
    
    void helper(string &digits, string curr, int index){
        if(index==digits.size()){
            result.push_back(curr); 
            return ;
        }
        char c=digits[index];
        vector<char> &tmp=um[c];
        for(const auto& t:tmp){
            helper(digits,curr+t,index+1);
        }
    }
};

File: 0017.letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
#https://www.cnblogs.com/kepuCS/p/5271654.html
#这道题目不算水，我做了好久，一开始思路错了，要想到用回溯。
#http://lib.csdn.net/article/datastructure/16705迭代解法也很清晰。
class Solution(object):
    dic={u'2':['a','b','c'],u'3':['d','e','f'],u'4':['g','h','i'],u'5':['j','k','l'],u'6':['m','n','o'],u'7':['p','q','r','s'],u'8':['t','u','v'],u'9':['w','x','y','z']}
    def recuive_combine(self,index,cur,digits,result):
        if index==len(digits) and len(cur)!=0:
            result.append(cur)
            return
        v=self.dic[digits[index]]
        for vv in v:
            self.recuive_combine(index+1,cur+vv,digits,result)
            
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        #digits=u'23'
        if len(digits)==0:
            return []
        result=[]
        self.recuive_combine(0,'',digits,result)
        return result
                
    
        
        
        
        
        
        
        
        
        
        
        
        